src/raft/raft.go

```diff
@@ -41,25 +46,344 @@ type ApplyMsg struct {
 // A Go object implementing a single Raft peer.
 //
 type Raft struct {
-	mu        sync.Mutex
+	mu          sync.RWMutex
+	role        uint64 //atomic
+	currentTerm int    //atomic counter
+
 	peers     []*labrpc.ClientEnd
 	persister *Persister
 	me        int // index into peers[]
 
+	majority int
+
+	electionTimeout  time.Duration
+	heartbeatTimeout time.Duration
+	leaderLease      time.Duration
+
+	heartBeatChan chan bool
+	lastContact   time.Time
+
+	killed bool
+	killCh chan bool
+
 	// Your data here.
 	// Look at the paper's Figure 2 for a description of what
 	// state a Raft server must maintain.
 
 }
 
+func (rf *Raft) isLeader() (isLeader bool) {
+	rf.mu.RLock()
+	defer rf.mu.RUnlock()
+	isLeader = rf.role == 2
+	return
+}
+
+func (rf *Raft) isCandidate() (isCandidate bool) {
+	rf.mu.RLock()
+	defer rf.mu.RUnlock()
+	isCandidate = rf.role == 1
+	return
+}
+
+func (rf *Raft) isFollower() (isFollower bool) {
+	rf.mu.RLock()
+	defer rf.mu.RUnlock()
+	isFollower = rf.role == 0
+	return
+}
+
+func (rf *Raft) makeLeader() {
+	rf.mu.Lock()
+	defer rf.mu.Unlock()
+	rf.role = 2
+	return
+}
+
+func (rf *Raft) makeCandidate() {
+	rf.mu.Lock()
+	defer rf.mu.Unlock()
+	rf.role = 1
+	return
+}
+
+func (rf *Raft) makeFollower() {
+	rf.mu.Lock()
+	defer rf.mu.Unlock()
+	rf.role = 0
+	return
+}
+
+func (rf *Raft) incTerm() {
+	rf.mu.Lock()
+	defer rf.mu.Unlock()
+	rf.currentTerm++
+	return
+}
+
+func (rf *Raft) decTerm() {
+	rf.mu.Lock()
+	defer rf.mu.Unlock()
+	rf.currentTerm--
+	return
+}
+
+func (rf *Raft) setTerm(term int) {
+	rf.mu.Lock()
+	defer rf.mu.Unlock()
+	rf.currentTerm = term
+	return
+}
+
+func (rf *Raft) getMajority() int {
+	rf.mu.RLock()
+	defer rf.mu.RUnlock()
+	return rf.majority
+
+}
+
+func (rf *Raft) setLastContact() {
+	rf.mu.Lock()
+	defer rf.mu.Unlock()
+	rf.lastContact = time.Now()
+}
+
+func (rf *Raft) LastContact() time.Time {
+	rf.mu.RLock()
+	defer rf.mu.RUnlock()
+	contact := rf.lastContact
+	return contact
+}
+
+func (rf *Raft) getID() (ID int) {
+	rf.mu.RLock()
+	defer rf.mu.RUnlock()
+	ID = rf.me
+	return
+}
+
+func (rf *Raft) getTerm() (term int) {
+	rf.mu.RLock()
+	defer rf.mu.RUnlock()
+	term = rf.currentTerm
+	return
+}
+
 // return currentTerm and whether this server
 // believes it is the leader.
 func (rf *Raft) GetState() (int, bool) {
-
-	var term int
-	var isleader bool
+	term := rf.getTerm()
+	isLeader := rf.isLeader()
 	// Your code here.
-	return term, isleader
+	return term, isLeader
+}
+
+// This is our main loop. It checks the state of the peer and drops down into the applicable loop
+func (rf *Raft) run() {
+	for {
+		switch {
+		case rf.killed:
+			DPrintf("Shutdown  🔫 ")
+			return
+		case rf.isFollower():
+			rf.runFollower()
+		case rf.isCandidate():
+			DPrintf("✋ Candidate state hit on %v", rf.me)
+			rf.runCandidate()
+		case rf.isLeader():
+			DPrintf("💪 leader state hit on %v", rf.me)
+			rf.runLeader()
+		}
+	}
+}
```

Comment:
This is our main loop. Based on the state of the given raft member, we'll drop down into subloops.